### PR TITLE
Extensive changes in arrays.scad, vectors.scad, common.scad and their regression test codes

### PR DESCRIPTION
**Pull request by RonaldoCMP**
Here is an extensive review of common.scad, vectors.scad, arrays.scad and their regression test codes. The changes were mostly in input data check and format. Some few functions got more changes either to adapt them to the new asserts or because a better (faster, simpler) code was found. 

Another change was moving add_scalar from vectors.scad to arrays.scad. It is in fact an array operation. Some few documentation examples and test codes were added. Some function argument names were changed to keep them consistent with the names they have in other similar functions. That name changes were not as extensive as I would like but it is advisable to favor users.

Most, but not all, changed function codes are followed by a comment with a short summary of the changes done. Those comments always start with:
//***

I can eliminate those comments in a second round.

### CODE REVIEW

File: vectors.scad [hunk 1]
```diff
@@ -19,41 +19,26 @@
 // Arguments:
 //   v = The value to test to see if it is a vector.
 //   length = If given, make sure the vector is `length` items long.
-//   zero = If false, require that the length of the vector is not approximately zero.  If true, require the length of the vector to be approx zero-length.  Default: `undef` (don't check vector length.)
```
[the remark this answers is not in the document]
Comment (RonaldoCMP) [reply vectors.scad:22]: You are correct. my fault.

File: vectors.scad [hunk 2]
```diff
@@ -76,7 +62,15 @@ function vang(v) =
...
 // Example:
 //   vmul([3,4,5], [8,7,6]);  // Returns [24, 28, 30]
-function vmul(v1, v2) = [for (i = [0:1:len(v1)-1]) v1[i]*v2[i]];
+function vmul(v1, v2) = 
+//    assert( is_vector(v1) && is_vector(v2,len(v1)), "Incompatible vectors")
+    assert( is_vector(v1) && is_vector(v2), "Invalid vector(s)")
+    [for (i = [0:1:len(v1)-1]) v1[i]*v2[i]];
+    
+//***
+// some other functions seem to rely on the multiplication of vectors with different lengths
```
[the remark this answers is not in the document]
Comment (RonaldoCMP) [reply vectors.scad:71]: Me too but there are at least two places in the library that don't please the regression tests if dissimilar vectors generate an error by `vmul`. That is why I couldn't change `vmul`.

File: arrays.scad [hunk 3]
```diff
@@ -44,58 +44,77 @@
...
-            [for (i = [s:1:e]) list[i]] :
-            concat([for (i = [s:1:l-1]) list[i]], [for (i = [0:1:e]) list[i]])
-    );
+    l==0 ? []
+    :   end==undef? 
+            is_num(start)?
+                list[ (start%l+l)%l ]
+            :   assert( is_list(start) || is_range(start), "Invalid start parameter")
+                [for (i=start) list[ (i%l+l)%l ] ]
+        :   assert(is_num(start), "Invalid start parameter.")
```
[the remark this answers is not in the document]
Comment (RonaldoCMP) [reply arrays.scad:55]: Copy that.

File: arrays.scad [hunk 4]
```diff
@@ -1201,15 +1324,18 @@ function _array_dim_recurse(v) =
...
+    (depth == undef)
+    ?   concat([len(v)], _array_dim_recurse(v))
+    :   (depth == 0)
+        ?  len(v)
+        :  let( dimlist = _array_dim_recurse(v))
+           (depth > len(dimlist))? 0 : dimlist[depth-1] ;
 
+//***
+// format
+// This function may return undef!
```
[the remark this answers is not in the document]
Comment (RonaldoCMP) [reply arrays.scad:1338]: Ok, but then the documentation should explicitly remark that.

I will do the changes above (excepting vmul) and try again. That time I will remove my comments in the code.

File: vectors.scad (hunk 2 above)
[the remark this answers is not in the document]
Comment (revarbat) [reply vectors.scad:71]: File a new issue against `vmul()` and I'll work on it after merging this PR,

**Comment by RonaldoCMP** [comment]
I have already filed an issue on `vmul` 3 days ago.

File: arrays.scad [hunk 5]
```diff
@@ -44,58 +44,77 @@
...
-//   l = The list to search.
-//   idx = If given, searches the given subindexes for matches for `x`.
+//   val = The simple value to search for.
+//   list = The list to search.
+//   idx = If given, searches the given subindexes for matches for `val`.
 // Example:
 //   in_list("bar", ["foo", "bar", "baz"]);  // Returns true.
 //   in_list("bee", ["foo", "bar", "baz"]);  // Returns false.
 //   in_list("bar", [[2,"foo"], [4,"bar"], [3,"baz"]], idx=1);  // Returns true.
+// Note:
```
Comment (revarbat) [inline arrays.scad:107]: Just noticed this... The docs generator doesn't know the `Note:` tag.  You'll need to put it at the end of the `Description:` block, indented to be part of that block.

I need to make the docs generator not silently ignore things it doesn't understand in a docs comment block.

Comment (RonaldoCMP) [reply arrays.scad:107]: Noted.

Have you seen the issue on vmul?

**Comment by revarbat** [comment]
Your latest commit completely deletes the polyhedra.scad, shapes.shad, and shapes2d.scad files.  Not good. Need any help with git commands?

**Comment by revarbat** [comment]
Here's a tutorial for removing a mistaken commit:
    https://samwize.com/2014/01/15/how-to-remove-a-commit-that-is-already-pushed-to-github/

**Comment by revarbat** [comment]
Okay, since you're using GitHub Desktop, you should be able to restore those files like this:
- Click on "History"
- Right-click on the commit for "restore polyhedra, shapes and shapes3d"
- Select "Revert this commit"
- Click on "Push origin"

**Pull request merged**